src/Numerics/SystemSolvers/preconditioners.jl

```diff
@@ -0,0 +1,134 @@
+export AbstractPreconditioner,
+    ColumnwiseLUPreconditioner,
+    NoPreconditioner,
+    preconditioner_update!,
+    preconditioner_solve!,
+    preconditioner_counter_update!
+
+"""
+    Abstract base type for all preconditioners.
+"""
+abstract type AbstractPreconditioner end
+
+"""
+mutable struct NoPreconditioner
+end
+
+Do nothing
+"""
+mutable struct NoPreconditioner <: AbstractPreconditioner end
+
+"""
+Do nothing, when there is no preconditioner, preconditioner = Nothing
+"""
+function preconditioner_update!(
+    op,
+    dg,
+    preconditioner::NoPreconditioner,
+    args...,
+) end
+
+"""
+Do nothing, when there is no preconditioner, preconditioner = Nothing
+"""
+function preconditioner_solve!(preconditioner::NoPreconditioner, Q) end
+
+"""
+Do nothing, when there is no preconditioner, preconditioner = Nothing
+"""
+function preconditioner_counter_update!(preconditioner::NoPreconditioner) end
+
+"""
+mutable struct ColumnwiseLUPreconditioner{AT}
+    A::DGColumnBandedMatrix
+    Q::AT
+    PQ::AT
+    counter::Int64
+    update_freq::Int64
+end
+
+A: the lu factor of the precondition (approximated Jacobian), in the DGColumnBandedMatrix format
+Q: MPIArray container, used to update A
+PQ: MPIArray container, used to update A
+counter: count the number of Newton, when counter > update_freq or counter < 0, update precondition
+update_freq: preconditioner update frequency
+
+"""
+mutable struct ColumnwiseLUPreconditioner{AT} <: AbstractPreconditioner
+    A::DGColumnBandedMatrix
+    Q::AT
+    PQ::AT
+    counter::Int64
+    update_freq::Int64
+end
+
+"""
+ColumnwiseLUPreconditioner constructor
+build an empty ColumnwiseLUPreconditioner
+
+dg: DG model, use only the grid information
+Q0: MPIArray, use only its size
+counter = -1, which indicates the preconditioner is empty
+update_freq: preconditioner update frequency
+
+"""
+function ColumnwiseLUPreconditioner(dg, Q0, update_freq = 100)
+    single_column = false
```

Comment:
Should we expose this? Or does that not make sense for `ColumnwiseLUPreconditioner`?

Reply:
This is one of the things I wanted to do. I think it should be exposed. But apparently people want this in fast... So I will wait to do this in a later PR.